modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1175,18 +1185,40 @@ def _take_2d_positional(
             all_rows = None
             if self.has_materialized_index:
                 all_rows = len(self.index)
-            elif self._row_lengths_cache:
+            elif self._row_lengths_cache or must_sort_row_pos:
                 all_rows = sum(self._row_lengths_cache)
```

Comment:
Should we use `row_lengths` here?
```suggestion
                all_rows = sum(self.row_lengths)
```